### PR TITLE
Allow inheritance in clades.tsv

### LINKED ISSUE #823: Inherited clade definitions
Currently clades are defined independently of one another in the provided TSV, but we often duplicate the mutations of a parent clade. For example, 21L is a descendant of 21M so we have the following:

```
21M (Omicron) 	nuc 	23525 	T
21M (Omicron) 	nuc 	23599 	G
21L (Omicron) 	nuc 	23525 	T  ## mutation actually defines 21M
21L (Omicron) 	nuc 	23599 	G  ## mutation actually defines 21M
21L (Omicron) 	nuc 	24424 	T
```

We should allow clades to be inherited, e.g.:

```
21M (Omicron) 	nuc 	23525 	T
21M (Omicron) 	nuc 	23599 	G
21L (Omicron) 	clade 	21M (Omicron)
21L (Omicron) 	nuc 	24424 	T
```

There are a few considerations here:
* This introduces the potential for circular dependencies (A descended from B descended from A) which should be fatal errors.
* When both a parent and descendant clade are annotated on the same branch, the branch label should represent the descendant clade.
* Should multiple parent clades be allowed? Probably easiest to limit the current implementation to a single parent lineage.

**Related**

* [Slack conversation](https://bedfordlab.slack.com/archives/C01UN8SALPM/p1640727022006200)
* [List of `augur clade` oddities, including some bugs related to parsing the TSV file, which should be addressed before / as part of this issue](https://github.com/nextstrain/augur/issues/735)

**Possible solution**   

There seem to be two implementations available:
1. Expand the TSV upon parsing to replace the parent clade with the mutations of the parent clade
2. Only consider the subtree defined by the parent clade and then find the clade defined by the extra mutations (e.g. 24424T in the example above). 

I prefer solution 2, but I don't think the results will be different.

### PR DESCRIPTION
**Pull request by corneliusroemer**
### Description of proposed changes

This PR extends `read_in_clade_definitions` in `clades.py` to allow overwritable inheritance using the following syntax:
```
    clade      gene    site     alt
    Clade_1    ctpE    81       D
    Clade_2    nuc     30642    T
    Clade_3    nuc     444296   A
    Clade_3    S       1        P
    # Clade_4 inherits from Clade_3
    Clade_4    clade   Clade_3
    Clade_4    pks8    634      T
    # Inherited allele can be overwritten
    Clade_4    S       1        L
```

Significant simplification of `clades.tsv` using this syntax can be observed in this ncov PR: https://github.com/nextstrain/ncov/pull/855/files

It uses networkx to create a directed graph of clades, to traverse in topological order and check that the graph is a DAG.

Helpful errors are raised whenever the graph is not a proper DAG.

I also enable comments in clade tsv by adding the kwarg `comment='#'` in the `pd.read_csv`

The function is rigorously unit tested by `test_clades.py` with 100% coverage, including the raising of exceptions.

Note: It may be necessary to add networkx as a dependency, if it isn't one already.

### Related issue(s)  

Resolves #823 

### Testing

Clades.py does not have any integration tests yet. I added unit tests just for the `read_in_clade_definitions` function. But given that the change is very local, I don't expect anything to break.

### CODE REVIEW

**Comment by victorlin** [comment]
Tests are failing with `ModuleNotFoundError` – you'll want to add `networkx` here:

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/setup.py#L53-L61

**Comment by corneliusroemer** [comment]
Thanks Victor for pointing me to where I need to change. You can do it to if you like, it's a branch on ncov itself for everyone to edit.

I'd be particularly interested in feedback on testing: is there a neater way than creating a new TSV for every test case?

Would it make sense to construct the TSVs test files programmatically to be DRYer? Or is it just fine like this, as it's nice and easy to reason?

**Comment by tsibley** [comment]
> Would it make sense to construct the TSVs test files programmatically to be DRYer? Or is it just fine like this, as it's nice and easy to reason?

I think this is just fine and preferable over generation to me. It's easier to reason when you can look at the static file, and there's no harm in a handful of additional test data files.

**Comment by victorlin** [comment]
> I'd be particularly interested in feedback on testing: is there a neater way than creating a new TSV for every test case?

Not following any convention, but what I did for `test_filter_groupby.py` was create valid source data (pd.DataFrame) in a [pytest fixture](https://docs.pytest.org/en/6.2.x/fixture.html):

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/tests/test_filter_groupby.py#L5-L15

Then "break" it in individual unit tests like so:

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/tests/test_filter_groupby.py#L67-L68

Though this approach isn't directly usable by your code as-is since `read_in_clade_definitions` takes a file instead of DataFrame. You could refactor to have something like `read_in_clade_definitions_from_df` as an intermediate, which is then testable using my approach, but that might be too much for this PR.

**Comment by tsibley** [comment]
Alternatively, the TSVs are short enough that it would be reasonable IMO to inline them as strings (`"""…"""`) directly into the test file and then write them out to temp files at runtime for passing to `read_in_clade_definitions()`. (pytest offers some features for passing a ready-to-use temp file to your test functions.) This maintains the nice property @victorlin mentions of being able to see the input data right above the expected output. I would advocate against constructing a DataFrame directly, as then we're not testing the full function which in practice does actually read a TSV file.

**Review by huddlej — CHANGES_REQUESTED** [review]
This is really cool, @corneliusroemer. The implementation is clean and the tests are easy to follow. For this type of data, I prefer the standalone TSV files to the inline data in Python because I could easily load these data into pandas from the Python prompt and test out alternate approaches to data processing.

I made a couple of suggestions below that could make `read_in_clade_definitions` a little easier to read and maintain.

File: augur/clades.py [hunk 1]
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
...
+    root = 0
+    # For every clade, add edge from root as default
+    for clade in df.clade.unique():
+        G.add_edge(root, clade)
+    
+    for _, row in df[df.gene == 'clade'].iterrows():
+        try:
+            G.remove_edge(root, row.clade)
+        except nx.NetworkXError:
+            raise ValueError(f"Clade {row.clade} seems to inherit from more than one clade, that's not allowed")
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply augur/clades.py]: Fixed

File: augur/clades.py [hunk 2]
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
...
+        G.add_edge(row.site, row.clade)
+    
+    if not nx.is_directed_acyclic_graph(G):
+        raise ValueError(f"Clade definitions contain cycles {list(nx.simple_cycles(G))}")
+
+    for clade in islice(nx.topological_sort(G),1,None):
+        try:
+            clades[clade] = clades[next(G.predecessors(clade))].copy()
+        except StopIteration:
+            raise ValueError(f"Clade {next(G.successors(clade))} inherits from non-existent clade {clade}")
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply augur/clades.py]: Implemented

File: augur/clades.py [hunk 3]
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
...
+        for _, row in df[(df.clade == clade) & (df.gene != 'clade')].iterrows():
+            clades[clade][(row.gene, int(row.site)-1)] = row.alt
+    
+    # Convert items from dict of tuples to list of tuples
+    for clade in clades:
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply augur/clades.py]: Implemented

File: augur/clades.py [hunk 4]
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
...
+        clades[clade] = list(map(lambda k: k[0] + (k[1],),clades[clade].items()))
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply augur/clades.py]: Thanks, implemented

File: augur/clades.py [hunk 5]
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
         clade definitions as :code:`{clade_name:[(gene, site, allele),...]}`
     '''
 
-    clades = defaultdict(list)
-    df = pd.read_csv(clade_file, sep='\t' if clade_file.endswith('.tsv') else ',')
-    for index, row in df.iterrows():
-        allele = (row.gene, row.site-1, row.alt)
-        clades[row.clade].append(allele)
-    clades.default_factory = None
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply augur/clades.py:40]: Fixed

**Review by corneliusroemer — COMMENTED** [review]
I think I've resolved all of your change requests @huddlej

**Comment by corneliusroemer** [comment]
> We will need to include `networkx` as a dependency in our Bioconda recipe, too.

Yes, this is important. If we don't yet do it, it'd be good to run proper tests on our Bioconda recipe too, so we would catch things like this automatically in case we happen to forget.

**Comment by corneliusroemer** [comment]
I've added a simple functional cram test using data from Zika. I removed all the sequences contained in the node data jsons to reduce impact on repo size.

File: augur/clades.py [hunk 6]
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
...
+        G.add_edge(root, clade)
+    
+    for _, row in clade_inheritance_rows.iterrows():
+        G.remove_edge(root, row.clade)
+        G.add_edge(row.site, row.clade)
+    
+    if not nx.is_directed_acyclic_graph(G):
+        raise ValueError(f"Clade definitions contain cycles {list(nx.simple_cycles(G))}")
+
+    for clade in islice(nx.topological_sort(G),1,None):
```
Comment (rneher) [inline augur/clades.py:78]: a few comments that explain what these functions do would be helpful. This is some sort of traversal, I guess, but can't tell from the code what kind.

File: augur/clades.py [hunk 7]
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
...
+        clades[clade] = clades[next(G.predecessors(clade))].copy()
```
Comment (rneher) [inline augur/clades.py:79]: same here. this is the parent?

Comment (rneher) [reply augur/clades.py:79]: and it adds definitions from parent clade to daughter clade

File: augur/clades.py [hunk 8]
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
...
+        for _, row in df[(df.clade == clade) & (df.gene != 'clade')].iterrows():
+            clades[clade][(row.gene, int(row.site)-1)] = row.alt
+    
+    # Convert items from dict[str, dict[(str,int),str]] to dict[str, list[(str,int,str)]]
+    clades = {
```
Comment (rneher) [inline augur/clades.py:84]: Maybe point out that mutations defined for child overwrite inherited definitions.

**Review by rneher — CHANGES_REQUESTED** [review]

File: augur/clades.py (hunk 6 above)
Comment (corneliusroemer) [reply augur/clades.py:91]: Do the comments I added help?

**Review by rneher — APPROVED** [review]

**Review by huddlej — APPROVED** [review]

**Comment by corneliusroemer** [comment]
@victorlin did you manually add networkx to the bioconda recipe or did the bioconda bump bot add the new dependency automatically?

It's definitely there which surprised me, thought I'd had to do it manually:
image

https://anaconda.org/bioconda/augur/files

**Comment by tsibley** [comment]
@corneliusroemer Manually: https://github.com/bioconda/bioconda-recipes/commit/14321554a95aa87309ee317ae218cce1de8f45f1

**Comment by victorlin** [comment]
Yeah, I caught the auto-bump PR bioconda/bioconda-recipes#32842 and told them to close it in favor of the manual PR.

**Pull request merged**